sample-code/spring-app/src/test/java/com/sap/ai/sdk/app/services/OpenAiServiceV2.java

```diff
@@ -110,6 +118,59 @@ public OpenAiChatCompletionResponse chatCompletionTools(final int months) {
     return OpenAiClient.forModel(GPT_4O_MINI).chatCompletion(request);
   }
 
+  /**
+   * Executes a chat completion request to OpenAI with a tool that calculates the weather.
+   *
+   * @param location The location to get the weather for.
+   * @param unit The unit of temperature to use.
+   * @return The assistant message response.
+   */
+  public OpenAiChatCompletionResponse chatCompletionToolExecution(
+      @Nonnull final String location, @Nonnull final String unit) {
+
+    final var jsonSchemaGenerator = new JsonSchemaGenerator(JACKSON);
+    Map<String, Object> schemaMap;
+    try {
+      final var schema = jsonSchemaGenerator.generateSchema(WeatherMethod.Request.class);
+      schemaMap = JACKSON.convertValue(schema, new TypeReference<>() {});
+    } catch (JsonMappingException e) {
+      throw new IllegalArgumentException("Could not generate schema for WeatherMethod.Request", e);
+    }
+
+    final var function =
+        new FunctionObject()
+            .name("weather")
+            .description("Get the weather for the given location")
+            .parameters(schemaMap);
+    final var tool = new ChatCompletionTool().type(FUNCTION).function(function);
+
+    final var messages = new ArrayList<OpenAiMessage>();
+    messages.add(OpenAiMessage.user("What's the weather in %s in %s?".formatted(location, unit)));
+
+    final var request = new OpenAiChatCompletionRequest(messages).withTools(List.of(tool));
+    final var response = OpenAiClient.forModel(GPT_4O_MINI).chatCompletion(request);
+
+    final var assistantMessage = response.getMessage();
+    messages.add(assistantMessage);
+
+    final var functionCall = (OpenAiFunctionCall) assistantMessage.toolCalls().get(0);
```

Comment:
`instanceof` check would be nicer:
- no casting
- custom error if type is not OpenAiFunctionCall